### ISSUE TITLE
HTTPX: x-kaazing-handshake protocol - Empty WebSocket binary frame marks the end of extended handshake response

### ISSUE THREAD

**Issue by sanjay-saxena** [k3po/k3po#299]
`x-kaazing-handshake` protocol as defined in [HTTPX specification](https://github.com/sanjay-saxena/k3po/blob/develop/specification/httpx/SPEC.md) must include an empty WebSocket binary frame at the end of extended handshake response. The empty WebSocket binary frame from the server looks as shown below:

`0x82 0x00`

Since the server may break the extended handshake response into multiple WebSocket frames, an empty WebSocket binary frame is used to indicate the end of the extended handshake response.

This means we have to update the specification as well as the existing spec scripts to include the empty WebSocket binary frame.

**Comment by cmebarrow** [comment]
@jfallows WDYT? Do you think this is the right approach?

**Comment by sanjay-saxena** [comment]
Here is the [PR](https://github.com/k3po/k3po/pull/302).

**Comment by jfallows** [comment]
As far as I remember, the empty WebSocket frame was a bug in the original implementation and should not be specified if not functionally needed.

Can we remove the requirement for the empty binary frame from the spec?

**Comment by sanjay-saxena** [comment]
Since the extended handshake response can be split across multiple WebSocket binary frames, currently the Gateway indicates the end of extended handshake response to the clients using an empty WebSocket frame. For example, here is how the extended handshake response looks when Gateway needs to send sending a `401` to the client:

```
# Extended handshake response - WebSocket Frame1
0x82 0x41
"HTTP/1.1 200 OK\r\n"
"Content-Type: text/html\r\n"
"Content-Length: 194\r\n"
"\r\n"

# Extended handshake response - WebSocket Frame2
0x82 0x7e 0x00 0x82
"HTTP/1.1 401 Unauthorized\r\n"
"WWW-Authenticate: Application Basic realm="
0x22
"Kaazing WebSocket Gateway Demo"
0x22
"\r\n"
"Content-Type: text/html\r\n"
"\r\n"

# Extended handshake response - WebSocket Frame3
0x82 0x40
"<html><head></head><body><h1>401 Unauthorized</h1></body></html>"

# Extended handshake response - WebSocket Frame4
0x82 0x00
```

**Comment by jfallows** [comment]
Thanks for the very clear explanation @sanjay-saxena. :smile: 

Wouldn't it make sense just to follow the `Content-Length` header value to determine when all bytes representing the response have been processed?

That would seem to make the empty frame redundant and the extended handshake response would be free to fragment across multiple WebSocket messages if necessary, with the recommended server behavior to send the extended handshake response in a single WebSocket message.

In the example above, the _outer_ response has a `Content-Length` header with value `194` which spans the entire _inner_ response, both inner response headers and inner response body. This is why there is no need for a `Content-Length` header on the inner response headers as it can be calculated implicitly and unambiguously.

The difference in behavior would be that the empty frame would actually be delivered to the application as an empty WebSocket message.

Therefore for backwards compatibility, we should probably add a header to the extended handshake request to let the server know it should not send the empty binary frame.

**Comment by sanjay-saxena** [comment]
Sounds good, @jfallows. Please let me know if this makes sense:

0.  Update [x-kaazing-handshake specification] (https://github.com/k3po/k3po/blob/develop/specification/httpx/SPEC.md) to define the behavior both in the presence and the absence of the proposed header in the extended handshake request.
0. Update the existing specification scripts to include the empty binary frame as the extended handshake request does _not_ contain the proposed header.
0. Add new specification tests such that extended handshake request contains the proposed header and the extended handshake response does _not_ contain empty binary frame.

We also need to do the following to implement this in the Gateway and the Clients:

0. File a ticket to add support for the proposed header to the extended handshake header in the Gateway.
0. File tickets for individual clients to send the proposed header in the extended handshake request and remove the requirement for an empty WebSocket frame to end the extended handshake response.

**Comment by jfallows** [comment]
I was thinking we should just specify the behavior we want it to be in the specification (with no empty binary frame) and then treat any remaining cases as separate backwards compatibility checks on the implementation, independent of the `x-kaazing-handshake` specification.

**Comment by sanjay-saxena** [comment]
Unless we capture the backward compatible behavior in some document(maybe separate from `x-kaazing-handshake` specification) and have specification scripts to test, we can very easily overlook/forget about it and may end up breaking older clients.

**Comment by sanjay-saxena** [comment]
@jfallows and I talked about this and here is where we have landed:

For the existing version of `x-kaazing-handshake` protocol, the spec will mandate requiring an empty binary frame to end the extended handshake response. In future, when we simplify the handshake process to avoid an additional round-trip to negotiate application extensions, we are going to introduce a new protocol which will obviate the need for the empty frame. Newer clients will negotiate the new protocol and will be able to use the `Content-Length` header in the WebSocket frame that arrives after the opening handshake response to figure out the negotiated extensions. Existing clients that use `x-kaazing-handshake` protocol will continue to receive the empty binary frame.
